File: Data Structures & Algorithms/longest-common-prefix/submission-3.py
class Solution:
    def longestCommonPrefix(self, strs: List[str]) -> str:
        if len(strs) == 0:
            return ""
        if len(strs) == 1:
            return strs[0]
        s1 = strs[0]
        s2 = strs[1]
        longest = 0
        while longest < min(len(s1), len(s2)) and s1[longest] == s2[longest]:
            longest += 1
        print(longest)
        for s in strs[2:]:
            if len(s) < longest:
                longest = len(s)
            while longest > 0 and s[longest-1] != s1[longest-1]:
                longest -= 1
        return s1[:longest] if longest > 0 else ""
         

File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-3.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int longest = 0;
        int lefti = 0;
        int righti = 0;
        Map<Character, Number> seen = new HashMap<>();
        char[] str = s.toCharArray();
        for(int i = 0; i < str.length; ++i){
            char c = str[i];
            if(seen.containsKey(c)){
                System.out.println("goobye key: " + c);
                System.out.println(seen);
                int currenti = lefti;
                while(str[currenti] != c){
                    seen.remove(str[currenti]);
                    currenti += 1;
                }
                lefti = currenti+1;
            }
            else{
                seen.put(c, 1);
                righti += 1;
                longest = longest < seen.size() ? seen.size() : longest;
            }
        }
        return longest;
    }
}
